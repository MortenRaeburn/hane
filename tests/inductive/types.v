Inductive prod (A : Set) (B : Set) : Set :=
    pair : forall (x : A) (y : B), prod A B.

Definition fst (A : Set) (B : Set) (p : prod A B) : A :=
    match p as p in prod _ _ return A with
    | pair _ _ x y => x
    end.

Definition snd (A : Set) (B : Set) (p : prod A B) : B :=
    match p as p in prod _ _ return B with
    | pair _ _ x y => y
    end.

Inductive sum (A : Set) (B : Set) : Set :=
    | inl : forall x : A, sum A B
    | inr : forall x : B, sum A B.

Inductive sig (A : Set) (P : forall x : A, Prop) : Set :=
    exist : forall (x : A) (p : P x), sig A P.

Definition proj1_sig (A : Set) (P : forall x : A, Prop) (x : sig A P) : A :=
    match x as _ in sig _ _ return A with
    | exist _ _ x _ => x
    end.

Definition proj2_sig (A : Set) (P : forall x : A, Prop) (x : sig A P) : P (proj1_sig A P x) :=
    match x as x in sig _ _ return P (proj1_sig A P x) with
    | exist _ _ _ p => p
    end.
