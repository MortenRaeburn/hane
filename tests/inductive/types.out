Inductive prod (Set) (Set) : Set :=
    | pair : forall['1] (forall['1] (((prod) ('3)) ('2))).
Definition fst : forall[Set] (forall[Set] (forall[((prod) ('1)) ('0)] ('2))) := fun[Set] (fun[Set] (fun[((prod) ('1)) ('0)] (match '0 in prod return '5 with pair => '1 end))).
Definition snd : forall[Set] (forall[Set] (forall[((prod) ('1)) ('0)] ('1))) := fun[Set] (fun[Set] (fun[((prod) ('1)) ('0)] (match '0 in prod return '4 with pair => '0 end))).
Inductive sum (Set) (Set) : Set :=
    | inl : forall['1] (((sum) ('2)) ('1))
    | inr : forall['0] (((sum) ('2)) ('1)).
Inductive sig (Set) (forall['0] (Prop)) : Set :=
    | exist : forall['1] (forall[('1) ('0)] (((sig) ('3)) ('2))).
Definition proj1_sig : forall[Set] (forall[forall['0] (Prop)] (forall[((sig) ('1)) ('0)] ('2))) := fun[Set] (fun[forall['0] (Prop)] (fun[((sig) ('1)) ('0)] (match '0 in sig return '5 with exist => '1 end))).
Definition proj2_sig : forall[Set] (forall[forall['0] (Prop)] (forall[((sig) ('1)) ('0)] (('1) ((((proj1_sig) ('2)) ('1)) ('0))))) := fun[Set] (fun[forall['0] (Prop)] (fun[((sig) ('1)) ('0)] (match '0 in sig return ('4) ((((proj1_sig) ('5)) ('4)) ('0)) with exist => '0 end))).
